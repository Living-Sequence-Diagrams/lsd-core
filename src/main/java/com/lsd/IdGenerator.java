package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
