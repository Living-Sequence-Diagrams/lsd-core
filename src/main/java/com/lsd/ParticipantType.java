package com.lsd;

import com.lsd.report.model.Participant;
import com.lsd.diagram.ValidComponentName;

public enum ParticipantType {
    ACTOR("actor"),
    BOUNDARY("boundary"),
    COLLECTIONS("collections"),
    CONTROL("control"),
    DATABASE("database"),
    ENTITY("entity"),
    PARTICIPANT("participant"),
    QUEUE("queue");

    private final String type;

    ParticipantType(final String type) {
        this.type = type;
    }

    public Participant called(String name) {
        return new Participant(String.format("%s %s", type, ValidComponentName.of(name)));
    }

    public Participant called(String name, String alias) {
        return new Participant(String.format("%s %s as \"%s\"", type, ValidComponentName.of(name), alias));
    }
}
