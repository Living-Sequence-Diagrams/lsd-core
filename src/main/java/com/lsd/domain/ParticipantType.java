package com.lsd.domain;

import org.apache.commons.text.WordUtils;

public enum ParticipantType {
    ACTOR("actor"),
    BOUNDARY("boundary"),
    COLLECTIONS("collections"),
    CONTROL("control"),
    DATABASE("database"),
    ENTITY("entity"),
    PARTICIPANT("participant"),
    QUEUE("queue");

    private final String type;

    ParticipantType(final String type) {
        this.type = type;
    }

    public Participant called(String name) {
        return new Participant(String.format("%s %s", type, removeSpaces(name)));
    }

    public Participant called(String name, String alias) {
        return new Participant(String.format("%s %s as \"%s\"", type, removeSpaces(name), alias));
    }

    private String removeSpaces(String s) {
        return WordUtils.capitalizeFully(s)
                .replaceAll(" ", "");
    }
}
