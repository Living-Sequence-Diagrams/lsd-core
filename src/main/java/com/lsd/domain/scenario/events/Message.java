package com.lsd.domain.scenario.events;

import lombok.Builder;
import lombok.Value;

import static java.lang.String.format;

@Value
@Builder
public class Message implements Interaction {
    String id;
    String label;
    String from;
    String to;
    Object data;

    public String toMarkup() {
        return format("%s -> %s:[[#%s %s]]",
                getFrom(), getTo(), getId(), getLabel());
    }
}
