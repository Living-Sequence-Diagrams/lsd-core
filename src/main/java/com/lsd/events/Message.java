package com.lsd.events;

import com.lsd.report.model.DataHolder;
import lombok.Builder.Default;
import lombok.Data;
import lombok.experimental.SuperBuilder;
import org.apache.commons.text.StringSubstitutor;

import java.util.Map;

import static com.lsd.events.ArrowType.SOLID;

@Data
@SuperBuilder
public class Message implements DataHolder {
    private final String id;
    private final String from;
    private final String to;
    private final String label;

    @Default
    private final Object data = "";

    @Default
    private final String colour = "";

    @Default
    private final ArrowType arrowType = SOLID;

    public String toMarkup() {
        return StringSubstitutor.replace("${from} ${arrow} ${to}: <text fill=\"${colour}\">[[#${id} {${tooltip}} ${label}]]</text>", Map.of(
                "from", getFrom(),
                "to", getTo(),
                "id", getId(),
                "tooltip", getLabel().replaceAll("<\\$.*?>", ""),
                "label", abbreviatedLabel(),
                "colour", getColour(),
                "arrow", getArrowType().toMarkup(getColour())
        ));
    }

}
