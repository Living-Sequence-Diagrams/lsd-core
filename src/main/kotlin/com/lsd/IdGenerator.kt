package com.lsd

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    
    private val idSupplier: Supplier<String> = if (isDeterministic)
        Supplier { counter.incrementAndGet().toString() }
    else
        Supplier { randomUUID().toString().replace("-".toRegex(), "") }


    operator fun next(): String = idSupplier.get()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}