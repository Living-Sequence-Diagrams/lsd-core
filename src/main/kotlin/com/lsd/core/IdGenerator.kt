package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}