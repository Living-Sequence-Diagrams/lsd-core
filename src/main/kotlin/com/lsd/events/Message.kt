package com.lsd.events

import com.lsd.builders.MessageBuilder
import com.lsd.events.ArrowType.SOLID

data class Message(
    override val id: String,
    override val label: String,
    override val data: Any? = null,
    override val from: String,
    override val to: String,
    override val colour: String = "",
    override val arrowType: ArrowType = SOLID,
) : SequenceMessage {
    companion object {
        @JvmStatic
        fun builder(): MessageBuilder = MessageBuilder()
    }
}