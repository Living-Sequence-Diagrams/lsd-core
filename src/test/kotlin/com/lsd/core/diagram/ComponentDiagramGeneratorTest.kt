package com.lsd.core.diagram

import com.lsd.core.IdGenerator
import com.lsd.core.builders.MessageBuilder.Companion.messageBuilder
import com.lsd.core.domain.Message
import com.lsd.core.domain.ParticipantType.*
import org.approvaltests.Approvals
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class ComponentDiagramGeneratorTest {
    private val idGenerator = IdGenerator(isDeterministic = true)

    @Test
    fun removesDuplicateInteractions() {
        val sampleMessage = sampleMessage()
        val diagramGenerator = ComponentDiagramGenerator(
            idGenerator = idGenerator,
            events = listOf(sampleMessage, sampleMessage, sampleMessage),
            participants = emptyList()
        )

        Approvals.verify(diagramGenerator.diagram()?.uml)
    }

    @Test
    fun removesDuplicateParticipants() {
        val sampleParticipant = ACTOR.called("Nick")
        val diagramGenerator = ComponentDiagramGenerator(
            idGenerator,
            listOf(sampleMessage()),
            listOf(sampleParticipant, sampleParticipant)
        )

        Approvals.verify(diagramGenerator.diagram()?.uml)
    }

    @Test
    fun replacesParticipantKeywordWithComponent() {
        val sampleParticipant = PARTICIPANT.called("SomeService")
        val diagramGenerator = ComponentDiagramGenerator(
            idGenerator = idGenerator,
            events = listOf(sampleMessage()),
            participants = listOf(sampleParticipant)
        )

        Approvals.verify(diagramGenerator.diagram()?.uml)
    }

    @Test
    fun convertsToValidComponentNames() {
        val sampleParticipant = PARTICIPANT.called("some service / name")
        val diagramGenerator =
            ComponentDiagramGenerator(
                idGenerator = idGenerator,
                events = listOf(sampleMessage()),
                participants = listOf(sampleParticipant)
            )

        Approvals.verify(diagramGenerator.diagram()!!.uml)
    }

    @Test
    fun convertsToValidComponentNamesForFromAndTo() {
        val diagramGenerator = ComponentDiagramGenerator(
            idGenerator = idGenerator,
            events = listOf(
                messageBuilder()
                    .from("fixes from")
                    .to("fixes to")
                    .build()
            ), participants = emptyList()
        )

        Approvals.verify(diagramGenerator.diagram()?.uml)
    }

    @Test
    fun mapsMessageEventsToParticipants() {
        val messages = listOf(
            messageBuilder().from("A").to("B").build(),
            messageBuilder().from("B").to("B").build(),
            messageBuilder().from("C").to("B").build(),
            messageBuilder().from("D").to("B").build(),
        )

        val participants = listOf(
            DATABASE.called("D", "dan", colour = "red"),
            BOUNDARY.called("B", "ben"),
            PARTICIPANT.called("unused")
        )

        assertThat(participants.usedIn(messages))
            .containsExactlyInAnyOrder(
                DATABASE.called("D", "dan", colour = "red"),
                BOUNDARY.called("B", "ben"),
                PARTICIPANT.called("A"),
                PARTICIPANT.called("C")
            )
    }

    @Test
    fun participantTypeAndAliasCanBeOverriddenInParticipantsButNotEvents() {
        val fred1 = CONTROL.called("Fred")
        val fred2 = DATABASE.called("Fred", "Freddy", "purple")
        val participants = listOf(fred1, fred2)

        val messages = listOf(
            messageBuilder().from(PARTICIPANT.called("Jake")).to(fred1).build(),
            messageBuilder().from("Fred").to("Bettie").build(),
            messageBuilder().from(PARTICIPANT.called("Fred", "Fredster", "red")).to("Bettie").build()
        )
        
        assertThat(participants.usedIn(messages))
            .containsExactly(
                PARTICIPANT.called("Jake"),
                fred2,
                PARTICIPANT.called("Bettie"),
            )
    }

    private fun sampleMessage(): Message {
        return messageBuilder()
            .from("A")
            .to("B")
            .build()
    }
}

