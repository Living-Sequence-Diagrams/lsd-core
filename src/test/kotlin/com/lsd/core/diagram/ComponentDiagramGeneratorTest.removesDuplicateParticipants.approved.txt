@startuml
!pragma layout smetana
!theme plain
skinparam backgroundColor transparent

actor Nick


[A] --> B


remove @unlinked
@enduml